### PR TITLE
[TACHYON-590] Implement the wildcard support for shell cmds with 1 argument

### PR DESCRIPTION
**Pull request by tianyin**
Per our discussion, we add wildcard support for shell cmds in multiple batches to ease the code review.

This PR contains the implementation of wildcard support for cmds with 1 argument, including the following cmds:
{cat, ls, lsr, rm, rmr, tail, touch, fileinfo, location, report, pin, unpin, free, du}

while the following cmds does not support wildcards by their semantic:
{mkdir, touch, count}
(where wildcard is treated as a regular character)

### CODE REVIEW

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 1]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+    expectedOutput += "/testWildCards/foo/foobar2 with file id " 
+        + mTfs.getFileId(new TachyonURI("/testWildCards/foo/foobar2")) 
+        + " has reported been report lost.\n";
+    expectedOutput += "/testWildCards/bar/foobar3 with file id " 
+        + mTfs.getFileId(new TachyonURI("/testWildCards/bar/foobar3"))
+        + " has reported been report lost.\n";
+    Assert.assertEquals(expectedOutput, mOutput.toString());
+  }
+  
+  private String getLocationResultStr(String path, int fid, TachyonFS tfs) throws IOException {
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: NIT: What about putting this `private` method at the end of the class to put the `@Test` together?

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 2]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+    Assert.assertFalse(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar1")).isInMemory());
+    Assert.assertFalse(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar2")).isInMemory());
+    Assert.assertFalse(mTfs.getFile(new TachyonURI("/testWildCards/bar/foobar3")).isInMemory());
+    Assert.assertFalse(mTfs.getFile(new TachyonURI("/testWildCards/foobar4")).isInMemory());
+    
+    mFsShell.run(new String[] {"load", "/testWildCards/*/foo*"});
+    
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar1")).isInMemory());
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar2")).isInMemory());
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/bar/foobar3")).isInMemory());
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: Assert `/testWildCards/foobar4` is not in memory here?

File: shell/src/test/java/tachyon/shell/TFsShellTest.java (hunk 1 above)
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: `fid` can be gotten once we have `path` and `tfs`?

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 3]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+    Assert.assertEquals(res2.contains("/testWildCards/bar/foobar3"), true);
+    Assert.assertEquals(res2.contains("/testWildCards/foobar4"), false);
+  }
+  
+  @Test
+  public void reportWildcardTest() throws IOException {
+    TFsShellUtilsTest.resetTachyonFileHierarchy(mTfs, WriteType.MUST_CACHE);
+    
+    mFsShell.run(new String[] {"report", "/testWildCards/*/foo*"});
+    String expectedOutput = "";
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: Better to use `StringBuilder`.

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 4]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+    expectedOutput += "/testWildCards/foo/foobar1 with file id " 
+        + mTfs.getFileId(new TachyonURI("/testWildCards/foo/foobar1")) 
+        + " has reported been report lost.\n";
+    expectedOutput += "/testWildCards/foo/foobar2 with file id "
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: These output messages are of the same format, may have a utility method.

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 5]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+  }
+  
+  @Test
+  public void freeWildCardTest() throws IOException {
+    TFsShellUtilsTest.resetTachyonFileHierarchy(mTfs, WriteType.MUST_CACHE);
+    
+    TachyonConf tachyonConf = mLocalTachyonCluster.getMasterTachyonConf();
+    
+    int ret = mFsShell.run(new String[] {"free", "/testWild*/foo/*"});
+    CommonUtils.sleepMs(null, tachyonConf.getInt(Constants.WORKER_TO_MASTER_HEARTBEAT_INTERVAL_MS,
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: The `free` command will block, why we need to sleep here?

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 6]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+        Constants.SECOND_MS) * 2 + 10);
+    Assert.assertEquals(0, ret);
+    Assert.assertFalse(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar1")).isInMemory());
+    Assert.assertFalse(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar2")).isInMemory());
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/bar/foobar3")).isInMemory());
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/foobar4")).isInMemory());
+    
+    ret = mFsShell.run(new String[] {"free", "/testWild*/*/"});
+    CommonUtils.sleepMs(null, tachyonConf.getInt(Constants.WORKER_TO_MASTER_HEARTBEAT_INTERVAL_MS,
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: Why sleep here?

File: shell/src/test/java/tachyon/shell/TFsShellTest.java (hunk 4 above)
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: Or we can have a `String[]` containing the paths like `/testWildCards/foo/foobar1, /testWildCards/foo/foobar2....`, then iterate over the array and append output to `expectedOutput`, then we can get rid of the redundant codes.

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 7]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+  
+  @Test
+  public void lsWildcardTest() throws IOException {
+    TFsShellUtilsTest.resetTachyonFileHierarchy(mTfs, WriteType.MUST_CACHE);
+    
+    TachyonURI expf1 = new TachyonURI("/testWildCards/foo/foobar1");
+    TachyonURI expf2 = new TachyonURI("/testWildCards/foo/foobar2");
+    TachyonURI expf3 = new TachyonURI("/testWildCards/bar/foobar3");
+    
+    String expected = "";
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: `StringBuilder`?

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 8]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+            TFsShell.convertMsToDate(mTfs.getFile(expf1).getCreationTimeMs()), "In Memory",
+            expf1.getPath());
+    expected +=
+        String.format(format, FormatUtils.getSizeFromBytes(20),
+            TFsShell.convertMsToDate(mTfs.getFile(expf2).getCreationTimeMs()), "In Memory",
+            expf2.getPath());
+    expected +=
+        String.format(format, FormatUtils.getSizeFromBytes(30),
+            TFsShell.convertMsToDate(mTfs.getFile(expf3).getCreationTimeMs()), "In Memory",
+            expf3.getPath());
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: If we have an array of sizes and an array of paths, then these `exptected += ` can be put into a `for` loop.

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 9]
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
...
+    
+    mFsShell.run(new String[] {"ls", "/testWildCards/*/foo*"});
+    Assert.assertEquals(expected, mOutput.toString());
+    
+    
+    TachyonURI expf4 = new TachyonURI("/testWildCards/foobar4");
+    expected +=
```
Comment (uronce-cc) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: These `exptected += ` can also be put into a for loop.

**Comment by uronce-cc** [comment]
@tianyin Finally, we have `tfs load /data/*` to load all data into memory, I always want this command before. :+1:   Just left a few minor comments for reducing code redundancy.

**Comment by tianyin** [comment]
@uronce-cc thanks a lot, Chang! You are always so helpful!

I'll address the comments asap (not in town today...)

I'll remove the single quote. I'm not a fan of it. Just saw in previous code there were some place use it but some place does not. So I just want to make it consistent. 

For the ``sleep`` in ``free``, I followed the ``freeTest`` so perhaps it's not necessary at all...

**Comment by tianyin** [comment]
@uronce-cc I have addressed all your comments except the ``sleep`` one. The sleep is needed otherwise the test case cannot pass (as the original ``freeTest``). 

Take a look :-)

**Comment by uronce-cc** [comment]
+1 LGTM

**Comment by calvinjia** [comment]
LGTM

**Pull request merged**